2022/Q3/artifacts/PoS-tla/MC_Staking.tla

```diff
@@ -160,34 +171,136 @@ HighCoverage(trace) ==
+ *
+ * The logic is the following:
+ * When we slash a validator (reduce its voting power) at the end of an epoch, we move tokens from the posAccount to
+ * the slashPool. Thus, in the posAccount, we only have the bonded tokens that are withdrawable: this means, the bonded
+ * tokens minus the slashsable amount. If a user is able to withdraw tokens that were contributing to the voting 
+ * power of the misbehaving validator, then in a execution were all toekns are withdrawn, the posAccount should go
+ * negative.
+*)
+
+PoSAccountAlwaysPositive == 
+    posAccount >= 0
+
+\* Auxiliary functions for Invariant #3.
+
+bondsWithOpenEnd == LET
+                    \* @type: (Set(BOND), <<ADDR, ADDR>>) => Set(BOND);   
+                    F(set, pair) == set \union { bond \in bonded[pair[1], pair[2]]: bond.end = -1 }
+                    IN ApaFoldSet(F, {}, UserAddrs \X ValidatorAddrs)
+
+foldUnbonds == LET
+               \* @type: (Set(UNBOND), <<ADDR, ADDR>>) => Set(UNBOND);   
+               F(set, pair) == set \union unbonded[pair[1], pair[2]]
+               IN ApaFoldSet(F, {}, UserAddrs \X ValidatorAddrs)
+
+(* 
+ * Invariant #3
+ * First bullet from https://github.com/informalsystems/partnership-heliax/issues/25
+ *
+ * The amount that is self-bonded and delegated tokens is exactly equal to the amount that can be withdrawn back from PoS.
+ * In other words, if all the bonds are withdrawn, the PoS balance must be equal to zero.
+*)
+
+posAccountZero ==
+    bondsWithOpenEnd = {} /\ foldUnbonds = {} => posAccount = 0
+
+\* Auxiliary functions for Invariants #4, #5 and #6.            
+
+SumBondsUser(user, val) == LET
+                           \* @type: (Int, BOND) => Int;    
+                           F(sum, bond) == sum + bond.amount
+                           IN ApaFoldSet(F, 0, { b \in bonded[user, val]:
+                                                 /\ b.start <= epoch
+                                                 /\ b.end /= -1 => b.end > epoch })
+
+TotalSumBonds(val) == LET F(sum, user) == sum + SumBondsUser(user, val)
+                      IN ApaFoldSet(F, 0, UserAddrs)
+
+(* 
+ * Invariant #4
+ * The validator's voting power at a given epoch is greater or equal to the total amount of
+ * tokens delegated to that validator.
+*)
 
 VotingpowerDelagations ==
     \A val \in ValidatorAddrs:
-    ValDelegatedFold(UserAddrs, val) = bonded[1, val] *)
+    TotalSumBonds(val) >= totalDeltas[UnbondingLength, val]
 
-\* Invariant #2
-\* the user balance is always greater or equal to zero
+(*
+ * Invariant #5
+ * The validator's voting power at a given epoch is not equal to the total amount of
+ * tokens delegated to that validator due to slashing.
+ * The follwoing invariant is used to create a counterexample.
+*)
 
-BalanceAlwaysPositive == 
-    \A user \in UserAddrs: balanceOf[user] >= 0
+VotingpowerNotEqualDelagations ==
+    \A val \in ValidatorAddrs:
+    TotalSumBonds(val) = totalDeltas[UnbondingLength, val]
```

Comment:
Shouldn't you also require at least one slash in the invariant?

Reply:
Why should I? I am not sure I am following.